### ISSUE TITLE
Consider dropping libp2p dependency

### ISSUE THREAD

**Issue by silasdavis** [ipfs/go-ipfs-api#166]
I was surprised when swapping out a hand-rolled IPFS HTTP API layer fo go-ipfs-api that my vendored dependencies grew by about 39000 lines. This appears to be almost entirely down to the dependency on "github.com/libp2p/go-libp2p-peer" (also pulled in by github.com/libp2p/go-libp2p-metrics:

https://github.com/ipfs/go-ipfs-api/blob/master/pubsub.go#L11

Which is used in pubsub.go in two places:

```go
	From     peer.ID
...
	from, err := peer.IDFromBytes(r.From)
```

Now `peer.ID` is just an alias for string, and `IDFromBytes` is just:

```go
func IDFromBytes(b []byte) (ID, error) {
	if _, err := mh.Cast(b); err != nil {
		return ID(""), err
	}
	return ID(b), nil
}
```

We also depend on libp2p-metrics (which pulls in peer) for the following struct (which we use use to deserialise into and never read and I cannot find a non-test usage) used here https://github.com/ipfs/go-ipfs-api/blob/master/shell.go#L477-L481:
```go
type Stats struct {
	TotalIn  int64
	TotalOut int64
	RateIn   float64
	RateOut  float64
}
```

Where mh is the multihash library on which we already depend. So we are paying the cost of a significant extra dependency for the benefit of a wrapper function on a library we already have and named type that is just a string. Is this dependency really pulling its weight?

I don't have anything against libp2p but for the purposes of auditability and upgrade it is better to keep down dependencies where possible and I would have thought a wrapper of an API would have few dependencies (many deliberately seek to have none - I slightly question whether we could do without multihash that pulls in various non-standard hashing libraries, but I don't know the code well enough).

Dependency graph below:

![dep](https://user-images.githubusercontent.com/99715/53082389-a7990a00-34f4-11e9-8f37-18000b599e27.png)

**Comment by Stebalien** [comment]
So, this library is actually going to be replaced by https://github.com/ipfs/go-ipfs-http-client (check out the in-progress PR) which will make that go-libp2p-peer dependency look like nothing. The goal is to make possible to use either an embedded go-ipfs node (as a library) or an external one (as a daemon) so this will only ever get worse.

One solution might be to extract the manual RPC part of the logic into a minimal library so users easily build their _own_ shells. That would allow users to write code like:

https://github.com/ipfs/go-ipfs-api/blob/b87a51132cbcfc47da594e9f39f9edb4b720570e/shell.go#L261-L264

**Comment by hsanjuan** [comment]
Note that even though peer.ID is internally a string, it is actually a binary byte slice. There is nothing readable about the internal string value of peer.ID.

The real question here is why ipfs is not returning human-readable responses from the pubsub endpoint, properly encoding the Peer.IDs to their human readable form like it does in other endpoints. It's seems it's returning raw pubsub messages roughly converted to json:

`{"from":"EiCHtiJBTplIUOYPmQnISQfPFixsQzfTQUNjAfFFHX0s/g==","data":"aGVsbG8=","seqno":"FYUsUlPpr00=","topicIDs":["hey"]}`

So there's actually no way to interact with this API endpoint without a library that can understand that binary multihash first on the client side.

**Comment by Stebalien** [comment]
Yeah, we should probably fix that. IIRC, it's an historical issue that we haven't fixed for stability issues however, pubsub _is_ experimental.

**Comment by silasdavis** [comment]
Thanks for responses and thanks @Stebalien  for heads up on https://github.com/ipfs/go-ipfs-http-client. I had a look at https://github.com/ipfs/go-ipfs (with it's precisely zero dependencies).

We have a project https://github.com/monax/hoard for grant-based encrypted content-addressed object sharing with multiple storage backends. I'm quite excited about the idea of embedding go-ipfs in Hoard and handling pinning and so on directly from that service.

@hsanjuan my point was that you don't need the `peer.ID` type to do that - you already have multihash and all you need is `multihash.Cast`.

**Comment by Stebalien** [comment]
This is super-low priority and unlikely to be fixed, unfortunately. We should be able to reduce the dependency list by implementing https://github.com/multiformats/go-multihash/issues/78 eventually, but that likely wont' happen for a while.
